File: other-programming-languages/Lua/2024/2024-004/2024-004.lua
-- Read the grid from a file
function read_grid(file_path)
    local grid = {}
    for line in io.lines(file_path) do
        table.insert(grid, line)
    end
    return grid
end

-- Get the dimensions of the grid
function grid_dimensions(grid)
    local rows = #grid
    local cols = #grid[1]
    return rows, cols
end

-- Check if a word exists in a specific direction
function check_word(grid, target_word, x, y, dx, dy, rows, cols)
    for i = 1, #target_word do
        local nx = x + (i - 1) * dx
        local ny = y + (i - 1) * dy
        if nx < 1 or ny < 1 or nx > rows or ny > cols or grid[nx]:sub(ny, ny) ~= target_word:sub(i, i) then
            return false
        end
    end
    return true
end

-- Count all occurrences of the target word
function count_word_occurrences(grid, target_word)
    local rows, cols = grid_dimensions(grid)
    local directions = {{0, 1}, {1, 0}, {1, 1}, {1, -1}, {0, -1}, {-1, 0}, {-1, -1}, {-1, 1}}
    local count = 0

    for r = 1, rows do
        for c = 1, cols do
            for _, dir in ipairs(directions) do
                local dx, dy = dir[1], dir[2]
                if check_word(grid, target_word, r, c, dx, dy, rows, cols) then
                    count = count + 1
                end
            end
        end
    end
    return count
end

-- Count all X-MAS patterns
function count_xmas_patterns(grid)
    local rows, cols = grid_dimensions(grid)
    local count = 0

    for r = 2, rows - 1 do
        for c = 2, cols - 1 do
            local center = grid[r]:sub(c, c)
            local top_left = grid[r - 1]:sub(c - 1, c - 1)
            local top_right = grid[r - 1]:sub(c + 1, c + 1)
            local bottom_left = grid[r + 1]:sub(c - 1, c - 1)
            local bottom_right = grid[r + 1]:sub(c + 1, c + 1)

            if center == "A" then
                -- Check all valid X-MAS configurations
                if top_left == "M" and top_right == "S" and bottom_left == "M" and bottom_right == "S" then
                    count = count + 1
                elseif top_left == "S" and top_right == "M" and bottom_left == "S" and bottom_right == "M" then
                    count = count + 1
                elseif top_left == "M" and top_right == "M" and bottom_left == "S" and bottom_right == "S" then
                    count = count + 1
                elseif top_left == "S" and top_right == "S" and bottom_left == "M" and bottom_right == "M" then
                    count = count + 1
                end
            end
        end
    end
    return count
end

-- Main function
function main()
    local file_path = "/uploads/input.txt"
    local target_word = "XMAS"

    -- Read the grid
    local grid = read_grid(file_path)

    -- Count occurrences of the target word
    local word_count = count_word_occurrences(grid, target_word)
    print("Count of '" .. target_word .. "': " .. word_count)

    -- Count X-MAS patterns
    local xmas_patterns = count_xmas_patterns(grid)
    print("Total X-MAS patterns: " .. xmas_patterns)
end

-- Run the main function
main()

File: other-programming-languages/Typescript/2024/2024-004/2024-004.ts
import * as fs from "fs";

// Function to read the grid from the input file
function readGrid(filePath: string): string[] {
    try {
        const data = fs.readFileSync(filePath, "utf8");
        return data.split("\n").map(line => line.trim()).filter(line => line.length > 0);
    } catch (err) {
        console.error("Error reading file:", err);
        return [];
    }
}

// Function to check if the word exists in a given direction
function checkWord(
    grid: string[],
    x: number,
    y: number,
    dx: number,
    dy: number,
    word: string,
    rows: number,
    cols: number
): boolean {
    for (let i = 0; i < word.length; i++) {
        const nx = x + i * dx;
        const ny = y + i * dy;

        if (nx < 0 || ny < 0 || nx >= rows || ny >= cols || grid[nx][ny] !== word[i]) {
            return false;
        }
    }
    return true;
}

// Count occurrences of the word "XMAS" in all directions
function countXMAS(grid: string[]): number {
    const targetWord = "XMAS";
    const directions = [
        [0, 1],   // Right
        [1, 0],   // Down
        [1, 1],   // Diagonal-right-down
        [1, -1],  // Diagonal-left-down
        [0, -1],  // Left
        [-1, 0],  // Up
        [-1, -1], // Diagonal-left-up
        [-1, 1]   // Diagonal-right-up
    ];

    const rows = grid.length;
    const cols = grid[0].length;
    let count = 0;

    for (let r = 0; r < rows; r++) {
        for (let c = 0; c < cols; c++) {
            for (const [dx, dy] of directions) {
                if (checkWord(grid, r, c, dx, dy, targetWord, rows, cols)) {
                    count++;
                }
            }
        }
    }

    return count;
}

// Count all X-MAS patterns
function countAllXMASPatterns(grid: string[]): number {
    const rows = grid.length;
    const cols = grid[0].length;
    let count = 0;

    for (let r = 1; r < rows - 1; r++) {
        for (let c = 1; c < cols - 1; c++) {
            const center = grid[r][c];
            const topLeft = grid[r - 1][c - 1];
            const topRight = grid[r - 1][c + 1];
            const bottomLeft = grid[r + 1][c - 1];
            const bottomRight = grid[r + 1][c + 1];

            if (center === 'A') {
                // Pattern 1: M.S
                if (topLeft === 'M' && topRight === 'S' && bottomLeft === 'M' && bottomRight === 'S') count++;
                // Pattern 2: S.M
                if (topLeft === 'S' && topRight === 'M' && bottomLeft === 'S' && bottomRight === 'M') count++;
                // Pattern 3: M.M
                if (topLeft === 'M' && topRight === 'M' && bottomLeft === 'S' && bottomRight === 'S') count++;
                // Pattern 4: S.S
                if (topLeft === 'S' && topRight === 'S' && bottomLeft === 'M' && bottomRight === 'M') count++;
            }
        }
    }

    return count;
}

// Main function
function main() {
    const filePath = "\\\\vmware-host\\Shared Folders\\C\\advent-of-code-002\\input-files\\2024\\2024-004\\input.txt";
    const grid = readGrid(filePath);

    if (grid.length === 0) {
        console.error("Error: Grid is empty or invalid.");
        return;
    }

    const xmasCount = countXMAS(grid);
    console.log(`Count of XMAS: ${xmasCount}`);

    const xmasPatterns = countAllXMASPatterns(grid);
    console.log(`Total X-MAS patterns: ${xmasPatterns}`);
}

// Run the main function
main();
